File: solutions/python/triangle/triangle.py
def equilateral(sides):
    return is_triangle(sides) and len(set(sides)) == 1


def isosceles(sides):
    return is_triangle(sides) and len(set(sides)) < 3


def scalene(sides):
    return is_triangle(sides) and len(set(sides)) == 3


def is_triangle(sides):
    return has_non_zero_sides(*sides) and satisfies_inequality(sides)


def has_non_zero_sides(a, b, c):
    return a > 0 and b > 0 and c > 0


def satisfies_inequality(sides):
    [a, b, c] = sorted(sides)
    return a + b >= c

File: solutions/csharp/triangle/Triangle.cs
public static class Triangle
{
    public static bool IsScalene(double side1, double side2, double side3)
    {
        if (!IsTriangle(side1, side2, side3))
        {
            return false;
        }
        
        return side1 != side2 && side1 != side3 && side2 != side3;
    }

    public static bool IsIsosceles(double side1, double side2, double side3) 
    {
        if (!IsTriangle(side1, side2, side3))
        {
            return false;
        }
        
        return side1 - side2 == 0 || side1 - side3 == 0 || side2 - side3 == 0 || side2 - side3 == 0; 
    }

    public static bool IsEquilateral(double side1, double side2, double side3) 
    {
        if (!IsTriangle(side1, side2, side3))
        {
            return false;
        }
        
        return side1 == side2 && side2 == side3;
    }

    private static bool IsTriangle(double side1, double side2, double side3) =>
        HasPositiveSides(side1, side2, side3) && SatisfiesInequality(side1, side2, side3);

    private static bool HasPositiveSides(double side1, double side2, double side3) =>
        side1 >= 0 && side2 >= 0 && side3 >= 0;

    private static bool SatisfiesInequality(double side1, double side2, double side3)
    {
        List<double> sides = [side1, side2, side3];
        sides.Sort();

        return sides[2] < (sides[1] + sides[0]);
    }
}